python/cudf_polars/cudf_polars/experimental/select.py

```diff
@@ -0,0 +1,67 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Parallel Select Logic."""
+
+from __future__ import annotations
+
+import operator
+from functools import reduce
+from typing import TYPE_CHECKING
+
+from cudf_polars.dsl.ir import Select
+from cudf_polars.dsl.traversal import traversal
+from cudf_polars.experimental.base import PartitionInfo
+from cudf_polars.experimental.dispatch import lower_ir_node
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.parallel import LowerIRTransformer
+
+
+_PARTWISE = (
+    "Literal",
+    "LiteralColumn",
+    "Col",
+    "ColRef",
+    "BooleanFunction",
+    "StringFunction",
+    "TemporalFunction",
+    "Filter",
+    "Cast",
+    "Ternary",
+    "BinOp",
+    "UnaryFunction",
+)
+
+
+@lower_ir_node.register(Select)
+def _(
+    ir: Select, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # Lower children
+    children, _partition_info = zip(*(rec(c) for c in ir.children), strict=True)
+    partition_info = reduce(operator.or_, _partition_info)
+    new_node = ir.reconstruct(children)
+
+    # Search the expression graph for "complex" operations
+    for ne in ir.exprs:
+        for expr in traversal(ne.value):
+            if type(expr).__name__ not in _PARTWISE:
+                # TODO: Lower Scan node into something else if an
+                # aggregattion is required. We assume that anything
+                # left as a Scan is partitionwise or single-partition.
+                if any(partition_info[c].count > 1 for c in children):
+                    raise NotImplementedError(
+                        f"Expr {type(expr)} does not support multiple partitions."
+                    )
+                else:  # pragma: no cover
+                    partition_info[new_node] = PartitionInfo(count=1)
+                    return new_node, partition_info
```

Comment:
>This can't be right

Sorry - I don't think I understand this comment?

Reply:
Ah, I think I read things wrong, sorry. This logic is (please correct me if I am wrong):

Does there exist an expression which is _not_ partition wise? If so, and our children have more than one partition, raise, otherwise we have a single partition.

How about this reworking:

```python
child, partition_info = rec(ir.children[0])
pi = partition_info[child]
if pi.count > 1 and not all(is_partitionwise(expr) for expr in traversal(ir.exprs)):
    raise NotImplementedError(...)
new_node = ir.reconstruct(child)
partition_info[new_node] = pi
```

Would need traversal to be changed to:
```diff
diff --git a/python/cudf_polars/cudf_polars/dsl/traversal.py b/python/cudf_polars/cudf_polars/dsl/traversal.py
index b3248dae93..9c45a68812 100644
--- a/python/cudf_polars/cudf_polars/dsl/traversal.py
+++ b/python/cudf_polars/cudf_polars/dsl/traversal.py
@@ -10,7 +10,7 @@ from typing import TYPE_CHECKING, Any, Generic
 from cudf_polars.typing import U_contra, V_co
 
 if TYPE_CHECKING:
-    from collections.abc import Callable, Generator, Mapping, MutableMapping
+    from collections.abc import Callable, Generator, Mapping, MutableMapping, Sequence
 
     from cudf_polars.typing import GenericTransformer, NodeT
 
@@ -23,22 +23,22 @@ __all__: list[str] = [
 ]
 
 
-def traversal(node: NodeT) -> Generator[NodeT, None, None]:
+def traversal(nodes: Sequence[NodeT]) -> Generator[NodeT, None, None]:
     """
     Pre-order traversal of nodes in an expression.
 
     Parameters
     ----------
-    node
-        Root of expression to traverse.
+    nodes
+        Roots of expressions to traverse.
 
     Yields
     ------
-    Unique nodes in the expression, parent before child, children
+    Unique nodes in the expressions, parent before child, children
     in-order from left to right.
     """
-    seen = {node}
-    lifo = [node]
+    seen = set(nodes)
+    lifo = list(nodes)
 
     while lifo:
         node = lifo.pop()
```

Reply:
Yes, we could do something similar to this. Though, `ir.exprs` returns `tuple[expr.NamedExpr, ...]` which is not a `Node` sub-class.

Reply:
Hence the required change to `traversal` to take a sequence of nodes.

Reply:
Right, I was making the unnecessary point that it would need to be something like `traversal([ne.value for ne 
 in ir.exprs])` (even if traversal accepts a sequence of Nodes)

Reply:
Oh, I see what you mean